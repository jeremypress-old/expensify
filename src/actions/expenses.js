import uuid from 'uuid';
import database from '../firebase/firebase';

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData
        const expense = { description, note, amount, createdAt };

        return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
}

export const removeExpense = ({ id }) => ({
    type: 'REMOVE_EXPENSE',
    expense: {
        id
    }
});

export const startRemoveExpense = ({ id }) => {
    return (dispatch) => {
        return database.ref('expenses').child(id).remove().then((snapshot) => {
            dispatch(removeExpense({ id }));
        });
    };
};

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch) => {
        return database.ref('expenses').child(id).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// START_SET_EXPENSES
export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        console.log(uid);
        return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
            let expenses = [];
            const val = snapshot.forEach((item) => {
                expenses.push({
                    id: item.key,
                    ...item.val()
                });
            });

            dispatch(setExpenses(expenses));
        });
    };
}
