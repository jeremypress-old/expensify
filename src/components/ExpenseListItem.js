import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export const ExpenseListItem = ({ description, amount, createdAt, id, dispatch, history},) => (
    <div>
        <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
        </Link>
        <p>{amount} - {createdAt}</p>
    </div>
);


export default ExpenseListItem
