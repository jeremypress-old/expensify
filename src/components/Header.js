import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink activeClassName="is-active" exact={true} to="/">Go Home!</NavLink>
        <NavLink activeClassName="is-active" to="/create">add expense</NavLink>
        <NavLink activeClassName="is-active" to="/edit">edit expense</NavLink>
        <NavLink activeClassName="is-active" to="/help">help</NavLink>

    </header>
)

export default Header;
