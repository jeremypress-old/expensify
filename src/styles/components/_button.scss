.btn {
    background-color: $light-blue;
    border: none;
    color: white;
    display: inline-block;
    font-weight: 300;
    font-size: $font-size-large;
    line-height: 1;
    padding: $s-size;
    text-decoration: none;
}

.btn--link {
    background: none;
}

.btn--secondary {
    background-color: #888888;
}
