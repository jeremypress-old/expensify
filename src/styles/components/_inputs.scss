.text-input {
    border: 1px solid #CACCCB;
    height: 50px;
    padding: $s-size;
    font-size: $font-size-large;
    font-weight: 300;
}

.select-input {
    @extend .text-input;
}

.select-area {
    @extend .text-input;
    height: 10rem;
}
