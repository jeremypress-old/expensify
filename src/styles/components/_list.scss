.list-header {
    background: $v-light-grey;
    border: 1px solid darken($v-light-grey, 7%);
    color: $grey;
    display: flex;
    justify-content: space-between;
    padding: $s-size $m-size;
}

.list-body {
    margin-bottom: $m-size;
}

.list-item {
    display: flex;
    flex-direction: column;
    border: 1px solid darken($v-light-grey, 10%);
    border-top: none;
    color: $dark-grey;
    padding: $s-size;
    text-decoration: none;
    transition: background .3s ease;

    &:hover {
        background: $v-light-grey;
    }

    @media (min-width: $desktop-breakpoint) {
        align-items: center;
        flex-direction: row;
        justify-content: space-between;
        padding: $m-size;
    }
}

.list-item__message {
    align-items: center;
    color: $grey;
    justify-content: center;
    padding: $m-size;
    &:hover {
        background: none
    }
}

.list-item__title {
    margin: 0;
    word-break: break-all;
}

.list-item__sub-title {
    color: $grey;
    font-size: $font-size-small;
}

.list-item__data {
    margin: $s-size 0 0 0;
    @media(min-width: $desktop-breakpoint) {
        margin: 0;
        padding-left: $s-size;
    }
}
