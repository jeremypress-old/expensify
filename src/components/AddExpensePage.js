import React from 'react';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends React.Component {
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )

    }

    onSubmit = (expense) => {
        this.props.onSubmit(expense)
        this.props.history.push('/')
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        onSubmit: (expense) => dispatch(startAddExpense(expense))
    }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
