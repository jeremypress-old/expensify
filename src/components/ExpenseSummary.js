import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import ExpenseListItem from './ExpenseListItem'
import selectVisibleExpenses from '../selectors/expenses';
import selectExpensesSum from '../selectors/expenses-total';


export const ExpenseSummary = ({expensesCount, sum}) => (
    <div>
        {expensesCount > 0 &&
            <h1>
                Viewing {expensesCount} {expensesCount=== 1 ? 'expense' : 'expenses'} for a total of {numeral(sum / 100).format('$0,0.00')}
            </h1>
        }
    </div>
)

const mapStateToProps = (state) => {
    const visibleExpenses = selectVisibleExpenses(state.expenses, state.filters)
    return {
        expensesCount: visibleExpenses.length,
        sum: selectExpensesSum(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpenseSummary);
