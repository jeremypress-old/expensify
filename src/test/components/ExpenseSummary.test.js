import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';

test('should render nothing with no expenses passed in', () => {
    const wrapper = shallow(<ExpenseSummary />);
    expect(wrapper).toMatchSnapshot();
});

test('should render one expense correctly ', () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={1} sum={expenses[0].amount} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render multiple expenses correctly ', () => {
    const wrapper = shallow(<ExpenseSummary expensesCount={expenses.length} sum={expenses[0].amount + expenses[1].amount + expenses[2].amount}/>);
    expect(wrapper).toMatchSnapshot();
});
