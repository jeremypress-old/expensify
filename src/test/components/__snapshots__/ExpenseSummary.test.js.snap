// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render multiple expenses correctly  1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     for a total of 
    $1,141.95
  </h1>
</div>
`;

exports[`should render nothing with no expenses passed in 1`] = `<div />`;

exports[`should render one expense correctly  1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     for a total of 
    $1.95
  </h1>
</div>
`;
